File: with_elixir/lib/day4.ex
defmodule Day4 do
  def part1(content) do
    content
    |> Enum.with_index()
    |> Enum.flat_map(fn {row, x} ->
      String.graphemes(row)
      |> Enum.with_index()
      |> Enum.filter(fn {char, _} -> char == "X" end)
      |> Enum.map(fn {_, y} -> {x, y} end)
    end)
    |> Enum.flat_map(fn {x, y} -> check_xmas_words(content, x, y) end)
    |> Enum.count()
  end

  def part2(content) do
    content
    |> Enum.with_index()
    |> Enum.drop(1)
    |> Enum.take(length(content) - 2)
    |> Enum.flat_map(fn {row, x} ->
      String.graphemes(row)
      |> Enum.with_index()
      |> Enum.drop(1)
      |> Enum.take(String.length(row) - 2)
      |> Enum.filter(fn {char, _y} -> char == "A" end)
      |> Enum.map(fn {_char, y} -> {x, y} end)
    end)
    |> Enum.count(fn {x, y} -> check_xmas_patterns(content, {x, y}) end)
  end

  defp at(grid, x, y) do
    if x >= 0 and y >= 0 and x < length(grid) and y < String.length(Enum.at(grid, 0)) do
      grid
      |> Enum.at(x)
      |> String.at(y)
    end
  end

  defp check_xmas_words(content, x, y) do
    [
      {-1, -1},
      {-1, 0},
      {-1, 1},
      {0, -1},
      {0, 1},
      {1, -1},
      {1, 0},
      {1, 1}
    ]
    |> Enum.filter(fn {dx, dy} ->
      "XMAS"
      |> String.graphemes()
      |> Enum.with_index()
      |> Enum.all?(fn {char, i} ->
        at(content, x + dx * i, y + dy * i) == char
      end)
    end)
  end

  defp check_xmas_patterns(content, {x, y}) do
    [
      at(content, x - 1, y - 1),
      at(content, x + 1, y - 1),
      at(content, x - 1, y + 1),
      at(content, x + 1, y + 1)
    ] in [
      ["M", "M", "S", "S"],
      ["S", "S", "M", "M"],
      ["M", "S", "M", "S"],
      ["S", "M", "S", "M"]
    ]
  end
end

File: 2025/with_rust/src/day4.rs
pub fn part1(content: Vec<String>) -> usize {
    let grid: Vec<Vec<char>> = generate_grid(content);

    grid.iter()
        .enumerate()
        .map(|(row, line)| {
            line.iter()
                .enumerate()
                .filter(|(col, cell)| **cell == '@' && count_adjacent_rolls(&grid, row, *col) < 4)
                .count()
        })
        .sum()
}

pub fn part2(content: Vec<String>) -> usize {
    fn count_then_remove(mut grid: Vec<Vec<char>>) -> usize {
        let accessible: Vec<_> = (0..grid.len())
            .flat_map(|row| (0..grid[row].len()).map(move |col| (row, col)))
            .filter(|&(row, col)| {
                grid[row][col] == '@' && count_adjacent_rolls(&grid, row, col) < 4
            })
            .collect();
        if accessible.is_empty() {
            return 0;
        }
        for (row, col) in &accessible {
            grid[*row][*col] = 'x';
        }
        accessible.len() + count_then_remove(grid)
    }

    count_then_remove(generate_grid(content))
}

fn generate_grid(content: Vec<String>) -> Vec<Vec<char>> {
    content.iter().map(|line| line.chars().collect()).collect()
}

fn count_adjacent_rolls(grid: &[Vec<char>], row: usize, col: usize) -> usize {
    [
        (-1, -1),
        (-1, 0),
        (-1, 1),
        (0, -1),
        (0, 1),
        (1, -1),
        (1, 0),
        (1, 1),
    ]
    .iter()
    .filter(|(dr, dc)| {
        grid.get((row as isize + dr) as usize)
            .and_then(|r| r.get((col as isize + dc) as usize))
            .map_or(false, |&cell| cell == '@')
    })
    .count()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::reader;

    #[test]
    fn part1solution() {
        assert_eq!(part1(reader::read_input("day4sample")), 13);
    }

    #[test]
    fn part2solution() {
        assert_eq!(part2(reader::read_input("day4sample")), 43);
    }
}
